packages/@aws-cdk/core/lib/private/resolve.ts

```diff
@@ -154,6 +154,13 @@ export function resolve(obj: any, options: IResolveOptions): any {
     throw new Error('Trying to resolve() a Construct at ' + pathName);
   }
 
+  // respect JavaScript's toJSON(): if the object has a toJSON() method, call it
+  // and continue with the returned value.
+  if ('toJSON' in obj && typeof(obj.toJSON) === 'function') {
```

Comment:
This should not really be necessary and fills me with dread.

Reply:
Why does it fill you with dread?

Reply:
* This is a finicky part of the code, and this change will touch a lot of runtime cases
* Token classes we use have a `toJSON` method, but it doesn't do anything meaningful (on purpose, because you can't/shouldn't use it!)
* This `toJSON` hook is being used in `CloudFormation.toJsonString()` to do some black magic
* I conceptually don't know how I feel about this change. Is `resolve()`ing the same as json-ifying? Feels like they should be two separate operations.

Is this so we can write:

```ts
const json = new CfnJson(this, 'Json', /* ... */);

useValue(json);
// instead of
useValue(json.jsonString);
```
?

Reply:
Come to think of it... This code could just be doing what (for example) Sfn StateMachine is doing, or CW Dashboard is doing, right?

No need for any changes to the resolve protocol

Reply:
And no need for CfnJson either?

Reply:
My mental model is that `resolve()` is designed to take only data (or tokens) as input. If you pass an object with e.g. methods or prototype to it, it will likely not do what you want. To that end, `toJSON()` is the JavaScript way to convert an object to data and as such I think it actually makes sense for `resolve()` to respect it. 

I added this _after_ we check if the value is a token, so tokens are handled before we get here. In this point of "resolve", we basically assume that the value is a "pure" data object.

This change actually surfaced a minor bug in [one of our tests] which to resolved a `PolicyStatement` object and the result was not an actual statement.

The reason this was needed here is not for `CfnJson` but because of how [`PhysicalResourceIdReference`](https://github.com/aws/aws-cdk/blob/762c8102d9315546f154047399be36d83a684127/packages/%40aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts#L17) is implemented in `AwsCustomResource`. It's a strong-type which represents a reference to the resource name and there's some encoding implemented via `toJSON()` for it in the custom resource protocol so it's possible to just use it anywhere in AWS API requests to plug in the reference to the resource name.

[one of our tests]: https://github.com/aws/aws-cdk/pull/13074/files#diff-8f1241f652c5d43f439b55aac2ac8edc274e48e6ff8d4a9ea020a29de5bc45d1L130

Reply:
> It's a strong-type which represents a reference to the resource name and there's some encoding implemented via toJSON() for it in the custom resource protocol so it's possible to just use it anywhere in AWS API requests to plug in the reference to the resource name.

But effectively, implementing `IResolvable` for that class would do the same thing. Why don't we do that?

Reply:
I can implement IResolvable But I actually think this is moving us forward. 

The previous implementation used JSON.stringify and therefore used toJSON, and as I replaced it with stack.stringifyJson it broke. For me, it surfaced this gap in support in the resolve/stringifyJson path.

Reply:
Just to summarize: we decided to use `IResolvable` in this PR and in add support for `toJSON()` in `resolve()` together with some additional checks afterwards to protect against passing non-data objects to `resolve()`.